tests/mfx_dispatch/linux/mfx_dispatch_test_cases_libs.cpp

```diff
@@ -19,32 +19,42 @@
 // SOFTWARE.
 
 #include "mfx_dispatch_test_main.h"
-
-#include "gtest/gtest.h"
+#include "mfx_dispatch_test_mock_call_obj.h"
+#include <gtest/gtest.h>
+#include <gmock/gmock.h>
 #include <mfxvideo.h>
-#include <functional>
+#include <algorithm>
 
-using namespace std::placeholders;
 
 TEST_F(DispatcherTest, ShouldSucceedForSeeminglyGoodMockLibrary)
 {
+    MockCallObj& mock = *g_call_obj_ptr;
+
     ver = {{MFX_VERSION_MINOR, MFX_VERSION_MAJOR}};
-    par.emulated_api_version = ver;
+    mock.emulated_api_version = ver;
 
-    SetupGoodLib(par);
+    EXPECT_CALL(mock, dlopen).Times(AtLeast(1)).WillRepeatedly(Return(MOCK_DLOPEN_HANDLE));
+    EXPECT_CALL(mock, dlsym).Times(AtLeast(1)).WillRepeatedly(Invoke(&mock, &MockCallObj::EmulateAPI));
+    EXPECT_CALL(mock, MFXInitEx).Times(AtLeast(1)).WillRepeatedly(DoAll(SetArgPointee<1>(MOCK_SESSION_HANDLE), Return(MFX_ERR_NONE)));
+    EXPECT_CALL(mock, MFXQueryIMPL).Times(AtLeast(1)).WillRepeatedly(Return(MFX_ERR_NONE));
+    EXPECT_CALL(mock, MFXQueryVersion).Times(AtLeast(1)).WillRepeatedly(Invoke(&mock, &MockCallObj::ReturnEmulatedVersion));
 
     mfxStatus sts = MFXInit(impl, &ver, &session);
     ASSERT_EQ(sts, MFX_ERR_NONE);
 }
 
 TEST_F(DispatcherTest, ShouldFailIfNoLibraryIsFound)
 {
+    MockCallObj& mock = *g_call_obj_ptr;
+    EXPECT_CALL(mock, dlopen).Times(AtLeast(1));
     mfxStatus sts = MFXInit(impl, &ver, &session);
     ASSERT_EQ(sts, MFX_ERR_UNSUPPORTED);
 }
 
 TEST_F(DispatcherTest, ShouldEnumerateCorrectLibNames)
```

Comment:
It would be better to have separate test cases instead of using 'for' loop, otherwise u'll get too generic test failure message like '[ FAILED ] DispatcherTest.ShouldEnumerateCorrectLibNames' and u need to parse log to understand which 'impl_case' really failed MFX_IMPL_AUTO, MFX_IMPL_SOFTWARE/whatever. If u separate this big test into several ones - u'll get clear message like '[ FAILED ] DispatcherTest.CheckCorrectLibNamesForAuto'.

To avoid code duplication, [gtest] 'Typed Test' approach might be used. Something like that:
```
template <int T>
using impl_type = std::integral_constant<int, T>;

using impl_case_list = testing::Types<
    impl_type<MFX_IMPL_SOFTWARE>,
    impl_type<MFX_IMPL_HARDWARE>
    //... other impl
>;

//DispatcherTest has to be template
TYPED_TEST_CASE(DispatcherTest, impl_case_list);
TYPED_TEST(DispatcherTest, CheckCorrectLibNames)
{
    int constexpr impl = TypeParam::value;
    //... here the same code as in 'for (auto impl_case : impl_cases_list) ' loop
```

If some certain case failed (let it be MFX_IMPL_HARDWARE == 2), u'd get the following message:

> [  FAILED  ] DispatcherTest/1.CheckCorrectLibNames, where TypeParam = struct std::integral_constant<int,2>

That's also is not a perfect message but u get a hint what happened which allows u to avoid gazing into the whole log.
As an additional benefit, this allows u to have finer grained test filters -
`--gtest_filter=DispatcherTest/<#case>.CheckCorrectLibNames`